mlflow/sklearn/__init__.py

```diff
@@ -668,7 +674,50 @@ def fit_mlflow(self, func_name, *args, **kwargs):
             else:
                 try_mlflow_log(mlflow.log_metric, "training_score", training_score)
 
-        try_mlflow_log(log_model, self, artifact_path="model")
+        SAMPLE_ROWS = 5
+        fit_arg_names = _get_arg_names(self.fit)
+        X_var_name, y_var_name = fit_arg_names[:2]
+
+        try:
+            X_sample = _get_Xy(args, kwargs, X_var_name, y_var_name)[0][:SAMPLE_ROWS]
+        except Exception as e:
+            X_sample = None
+            _logger.warning(
+                "Failed to sample `X`. The model signature and input example will not be recorded: "
+                + str(e)
+            )
+
+        # A few estimators define `fit_predict`, but don't define `predict`:
+        # Example: https://scikit-learn.org/stable/modules/generated/sklearn.cluster.AgglomerativeClustering.html  # NOQA
+        # For models that don't define `predict`, use `fit_predict` to get the model prediction.
+        has_predict = hasattr(self, "predict") or hasattr(self, "fit_predict")
+
+        signature = None
+
+        if (X_sample is not None) and has_predict:
+            # If this function is `fit_predict`, reuse its output (which should be a numpy array).
+            if func_name == "fit_predict":
+                model_output = fit_output[:SAMPLE_ROWS]
+            else:
+                try:
+                    model_output = (
+                        self.predict(X_sample)
+                        if hasattr(self, "predict")
+                        # use `clone` to prevent re-fitting
+                        else clone(self).fit_predict(X_sample)
```

Comment:
Use `sklearn.base.clone` to prevent re-fitting if we have to use `fit_predict`.

Reply:
If `hasattr(self, "predict")` is False, then, based on the previous discussion with @tomasatdatabricks, we should not output the pyfunc flavor or a model signature. Accordingly, I think we can get rid of this `else` case.